File: providers/streaming_test.go
package providers

import (
	"encoding/json"
	"io"
	"strings"
	"testing"
)

func TestChoiceFinishReasonJSONShape(t *testing.T) {
	// intermediate chunk: nil FinishReason must serialize as null, not vanish.
	b, _ := json.Marshal(Choice{Index: 0, Delta: &Delta{}})
	if !strings.Contains(string(b), `"finish_reason":null`) {
		t.Errorf("intermediate choice = %s, want finish_reason:null", b)
	}

	// terminal chunk: populated FinishReason serializes as its value.
	fr := "stop"
	b, _ = json.Marshal(Choice{Index: 0, Delta: &Delta{}, FinishReason: &fr})
	if !strings.Contains(string(b), `"finish_reason":"stop"`) {
		t.Errorf("terminal choice = %s, want finish_reason:stop", b)
	}
}

func TestParseStreamChunkErrorEnvelope(t *testing.T) {
	// an error envelope must surface as an error, not an all-zero chunk.
	_, err := parseStreamChunk(`{"error":{"message":"rate limit exceeded","type":"rate_limit_error"}}`)
	if err == nil {
		t.Fatal("error envelope must return an error")
	}
	if !strings.Contains(err.Error(), "rate limit exceeded") {
		t.Errorf("err = %q, want the upstream message", err)
	}

	// a normal chunk parses as a chunk.
	chunk, err := parseStreamChunk(`{"id":"chatcmpl-1","object":"chat.completion.chunk"}`)
	if err != nil {
		t.Fatalf("normal chunk parse = %v, want nil", err)
	}
	if chunk == nil || chunk.ID != "chatcmpl-1" {
		t.Errorf("chunk = %+v, want id chatcmpl-1", chunk)
	}
}

func TestParseStreamChunkReasoningAndCost(t *testing.T) {
	chunk, err := parseStreamChunk(`{"choices":[{"index":0,"delta":{"reasoning":"thinking"}}],"usage":{"cost":0.001}}`)
	if err != nil {
		t.Fatal(err)
	}
	if chunk.Choices[0].Delta.Reasoning != "thinking" || chunk.Usage == nil || chunk.Usage.Cost == nil || *chunk.Usage.Cost != 0.001 {
		t.Fatalf("chunk = %+v", chunk)
	}
}

func TestOpenAIStreamErrorEnvelopeSurfaces(t *testing.T) {
	o := NewOpenAI("test-key", "")

	sse := strings.Join([]string{
		`data: {"id":"chatcmpl-1","object":"chat.completion.chunk","choices":[{"index":0,"delta":{"content":"partial"}}]}`,
		``,
		`data: {"error":{"message":"upstream overloaded","type":"server_error"}}`,
		``,
	}, "\n")

	events := make(chan StreamEvent, 10)
	go o.readSSEStream(io.NopCloser(strings.NewReader(sse)), events)

	var streamErr error
	var done bool
	for ev := range events {
		if ev.Err != nil {
			streamErr = ev.Err
		}
		if ev.Done {
			done = true
		}
	}

	if streamErr == nil {
		t.Fatal("mid-stream error envelope must surface as a stream error")
	}
	if !strings.Contains(streamErr.Error(), "upstream overloaded") {
		t.Errorf("stream error = %q, want the upstream message", streamErr)
	}
	if done {
		t.Error("an errored stream must not also emit Done")
	}
}

File: desktop/src/providers/streaming.test.ts
import test from 'node:test';
import assert from 'node:assert/strict';
import { parseStreamChunk } from './streaming.js';
import { OpenAiProvider } from './openai.js';
import { startTestServer } from '../test-support/server.js';
import type { Choice } from './types.js';

test('intermediate choice serializes finish_reason as null', () => {
  const intermediate: Choice = { index: 0, delta: {}, finish_reason: null };
  assert.match(JSON.stringify(intermediate), /"finish_reason":null/);

  const terminal: Choice = { index: 0, delta: {}, finish_reason: 'stop' };
  assert.match(JSON.stringify(terminal), /"finish_reason":"stop"/);
});

test('parseStreamChunk surfaces an error envelope', () => {
  assert.throws(
    () => parseStreamChunk('{"error":{"message":"rate limit exceeded","type":"rate_limit_error"}}'),
    /rate limit exceeded/,
  );

  const chunk = parseStreamChunk('{"id":"chatcmpl-1","object":"chat.completion.chunk"}');
  assert.equal(chunk.id, 'chatcmpl-1');
});

test('parseStreamChunk keeps reasoning and cost', () => {
  const chunk = parseStreamChunk(
    '{"choices":[{"index":0,"delta":{"reasoning":"thinking"}}],"usage":{"cost":0.001}}',
  );
  assert.equal(chunk.choices[0]?.delta?.reasoning, 'thinking');
  assert.equal(chunk.usage?.cost, 0.001);
});

test('a mid-stream error envelope surfaces and the stream does not complete normally', async () => {
  const sse = [
    'data: {"id":"chatcmpl-1","object":"chat.completion.chunk","choices":[{"index":0,"delta":{"content":"partial"}}]}',
    '',
    'data: {"error":{"message":"upstream overloaded","type":"server_error"}}',
    '',
  ].join('\n');

  const server = await startTestServer((_req, res) => {
    res.writeHead(200, { 'Content-Type': 'text/event-stream' });
    res.end(sse);
  });

  try {
    const provider = new OpenAiProvider({ apiKey: 'test-key', baseUrl: server.url });
    const received: string[] = [];
    let streamErr: unknown;

    try {
      for await (const chunk of provider.chatCompletionStream({ model: 'gpt-4', messages: [] })) {
        received.push(chunk.choices[0]?.delta?.content ?? '');
      }
      // reaching here means the stream ended cleanly, which the envelope forbids
      assert.fail('mid-stream error envelope must surface as a stream error');
    } catch (err) {
      streamErr = err;
    }

    assert.deepEqual(received, ['partial']);
    assert.match(String(streamErr), /upstream overloaded/);
  } finally {
    await server.close();
  }
});
